### PR TITLE
[ESQL] Make query wrapped by SingleValueQuery cacheable

### PR DESCRIPTION
**Pull request by iverase**
This commit simplifies SingleValueQuery by introducing a new lucene query called SingleValueMatchQuery. This query iterates over doc values and return only the fields that are single-valued. Then SingleValueQuery can generate a lucene query which is a boolean query with two filters, one with the original query and the other with a SingleValueMatchQuery.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueMatchQuery.java [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+            this.score = score;
+            this.scoreMode = scoreMode;
+            this.maxDoc = maxDoc;
+            this.matchCost = matchCost;
+            this.predicate = predicate;
+        }
+
+        @Override
+        public Scorer get(long leadCost) {
+            TwoPhaseIterator iterator = new TwoPhaseIterator(DocIdSetIterator.all(maxDoc)) {
```
Comment (iverase) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueMatchQuery.java:323]: I wonder if we can do something better here

**Comment by iverase** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueQueryTests.java [hunk 2]
```diff
@@ -403,18 +279,5 @@ public List<List<Object>> build(RandomIndexWriter iw) throws IOException {
...
-        public void assertStats(SingleValueQuery.Builder builder, YesNoSometimes subHasTwoPhase) {
-            assertThat(builder.stats().missingField(), equalTo(1));
-            assertThat(builder.stats().numericSingle(), equalTo(0));
-            assertThat(builder.stats().numericMultiNoApprox(), equalTo(0));
-            assertThat(builder.stats().numericMultiApprox(), equalTo(0));
-            assertThat(builder.stats().ordinalsSingle(), equalTo(0));
-            assertThat(builder.stats().ordinalsMultiNoApprox(), equalTo(0));
-            assertThat(builder.stats().ordinalsMultiApprox(), equalTo(0));
-            assertThat(builder.stats().bytesApprox(), equalTo(0));
-            assertThat(builder.stats().bytesNoApprox(), equalTo(0));
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueQueryTests.java:417]: I thought the stats would be useful in debugging slow queries. It's certainly nice to know things like "this rewrote itself to match all in 7/8 segments".

Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueQueryTests.java:417]: But I don't think it's a blocker.

**Review by nik9000 — APPROVED** [review]

**Comment by iverase** [comment]
@elasticmachine update branch

**Pull request merged**